Mysql support

Odebe opened the pull request:
### What is the purpose of this pull request?
To add mysql support to Logidze.

### What changes did you make? (overview)
#### Implemented
All main Logidze features (filtering, triggers, history limit, snapshots, check pendind, meta*, log-togglem fx-support and settings) are implemented.

Refactoring to decouple database-dependent code from regular code. 
I created `Logidze::Implementation` that sets on load `Logidze::Implementation::Current` constant to module with current database  implementation (`Logidze::Implementation::Pg` or `Logidze::Implementation::Mysql`). 
So code depends only on `Logidze::Implementation::Current`.

Generator tasks split into subtasks for each database, but also can be invoked the old way (subtask will be selected based on db adapter config).

Tested with MySQL 8 and MySQL 5.7

#### Not implemented
* Exceptions handling  **(backlog)**
* `--except` option for `model:install` **(by-design)**
* Debounce **(backlog)**
* Partition (after_trigger) **(backlog)**
* Timestamp field **(backlog)**
* *Meta works unstable with associations touching **(backlog)**

#### WIP
 --
 
#### Backlog
* Write test for Implementation interface and selection
* Move common from generators to helpers
* Update documentation

### Is there anything you'd like reviewers to focus on?

### Checklist

- [ ] I've added tests for this change
- [ ] I've added a Changelog entry
- [ ] I've updated a documentation (Readme)

Odebe committed:
generator refactoring

Odebe committed:
selection works fine

Odebe committed:
generates migration with blank procedures

Odebe committed:
passes install migration for mysql

Odebe committed:
[remove or rewrite later] added dip

Odebe committed:
functions and triggers implementation (passes all acceptance specs)

Odebe committed:
triggers specs passed

Odebe committed:
Logidze::Error and Logidze::NotImplemented added

Odebe committed:
function definitions and serialization fixes

Odebe committed:
removing fx test for mysql

Odebe committed:
sql specs directory structure refactoring

Odebe committed:
mysql logidze_compact_history fix

Odebe committed:
sql helper fix (mysql query result does not respond to #values

Odebe committed:
remove adapters (code moved to implementations)

Odebe committed:
disabling postgresql sql specs for mysql

Odebe committed:
logidze_filter_keys specs

Odebe committed:
shapshot and versions specs and fixes

Odebe committed:
meta mysql fixes

Odebe committed:
specs

Odebe committed:
removing except from mysql spec pool

Odebe committed:
trigger spec fix

Odebe committed:
CI config

Odebe committed:
rubocop

Odebe committed:
new implementation selection process

Odebe committed:
removing byebug

Odebe committed:
database.yml copying as additional step

Odebe committed:
actions fix

Odebe committed:
actions fix [2]

Odebe committed:
specific url per database type

Odebe committed:
skip postgres in CI (for now)

Odebe committed:
moving database.yml copying to dummy app

Odebe committed:
actions fix [3]

Odebe committed:
moved implementation loading to engine initializers

Odebe committed:
Update rspec.yml

Odebe committed:
Update rspec.yml

Odebe committed:
Update rspec.yml

Odebe committed:
dummy fixes

Odebe committed:
Update rspec.yml

Change localhost to 127.0.0.1 in MYSQL_URL

Odebe committed:
fileutils fix

Odebe committed:
removing generated migration from dummy

Odebe committed:
actions update

Odebe committed:
Update rspec.yml

Odebe committed:
Update rspec.yml

Odebe committed:
Update rspec.yml

Odebe committed:
database_type usage instead of adapter_name

Odebe committed:
selecting implementation without calling establish_connection

Odebe committed:
rubocop

Odebe committed:
mysql 5.7 in actions

Odebe committed:
mysql 5.7: remove json_value from functions

Odebe committed:
mysql 5.7: removing [last] from json path

Odebe committed:
mysql 5.7: removing [last] from json path [2]

Odebe committed:
mysql 5.7: safe version casting

Odebe committed:
postgres: snapshot implementation fix

Odebe committed:
abstract: setting implementation fix

Odebe committed:
postres: from_fs function defenitions fix

Odebe committed:
dummy: fx fix

Odebe committed:
enable postgres job in actions

Odebe committed:
removing dip

Odebe committed:
rubocop

Odebe committed:
Update rspec.yml

Odebe committed:
Update rspec.yml

Odebe committed:
Update rspec.yml

Odebe committed:
improvements in mysql logger function

Odebe committed:
mysql: clear setting in ruby instead of error handling

Odebe committed:
disabling unstable specs for mysql

Odebe committed:
rubocop

palkan commented on .github/workflows/rspec.yml:
```diff
@@ -7,14 +7,76 @@ on:
+      - name: Prepare test database user
+        run: |
+          mysql -h0.0.0.0 -uroot -proot -e "GRANT SUPER ON *.* TO rails; FLUSH PRIVILEGES;"
+      - name: Run RSpec (unit, integrations)
+        run: |
+          bundle exec rspec --exclude-pattern=spec/acceptance/**/* -f d --force-color
+      - name: Run RSpec (acceptance)
+        run: |
+          bundle exec rspec spec/acceptance/ -f d --force-color
```
Why do we separate these tests?

Odebe replied:
In order to get fast feedback. Unit and integration tests are more detailed so it easier to figure out what went wrong.
And after we know unit/integration tests are passed we can check user-level (acceptance) tests.
Also it saves some of CI time.

palkan commented on lib/logidze/engine.rb:
```diff
@@ -1,19 +1,24 @@
     end
 
     initializer "check Logidze function versions" do |app|
+      require "logidze/utils/check_pending"
```
Why did we move this require statement to initializer?

Odebe replied:
Because we have to load [implementation-dependent modules](https://github.com/Odebe/logidze/blob/f6b0dd3b5bbd6e5f753c9ce7f7d7bf98ca55033b/lib/logidze/utils/function_definitions.rb#L8) after rails setup (i. e. when rails have loaded database config). 

 I think I should refoctor `Logidze::Utils::FunctionDefinitions` users to call straight  to`Logidze::Implementation::Current::FunctionDefinitions`.

palkan commented on lib/logidze/implementation/load.rb:
```diff
@@ -0,0 +1,21 @@
+        raise "Not supported database '#{adapter_name}'"
+      end
+
+    const_set("Current", current)
```
Perfect 👍 

(The file should called `current.rb` though)

palkan commented on spec/dummy/config/database.postgresql.yml:
```diff
@@ -1,6 +1,6 @@
   adapter: postgresql
   encoding: unicode
-  url: <%= ENV['DATABASE_URL'] %>
+  url: <%= ENV['POSTGRESQL_URL'] %>
```
Let's leave `DATABASE_URL` as a fallback (and for mysql, too).

palkan commented on spec/dummy/db/mysql/20160415194001_create_posts.rb:
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+class CreatePosts < ActiveRecord::Migration[5.0]
```
I wonder if we can re-use the same migrations for both databases; is there anything we need to take into account beyond `jsonb` vs. `json`?

Odebe replied:
Besides jsonb we should take in account an arrays. MySQL doesn't have them.
I've replaced `t.string :settings, array: true` with `t.json :settings` in user migration.
(I don't count partitioned table migration because it's support not implemented for mysql in PR)

One naive way to re-use migrations is to create helper module and select field type in migrations like:
```ruby
if postgresql?
  t.string :settings, array: true
end

if mysql?
  t.json :settings
end
```

palkan reviewed (commented):
Thanks!

Great work; left some minor comments.

Probably, the only one that I would like to take care of is unifying test migrations.

Let's wait for reviews from others (until Wed), and then proceed.

ardecvz commented on lib/logidze/implementation/mysql.rb:
```diff
@@ -0,0 +1,13 @@
+
+module Logidze
+  module Implementation
+    module Mysl
```
A typo.

ardecvz commented on lib/logidze/implementation.rb:
```diff
@@ -0,0 +1,38 @@
+      def database_type
+        @database_name ||=
+          case adapter_name
+          when "mysql2"
```
Let's check for Trilogy too as it's included in Rails 7, and I believe you do not need to apply any changes to support it - https://github.com/rails/rails/pull/47880

ardecvz commented on lib/logidze/error.rb:
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+module Logidze
```
:+1: for this solid error hierarchy, but personally I do not understand CantDefineDatabase - maybe CannotIdentifyDatabase or something like this.

ardecvz commented on lib/generators/logidze/model/mysql/triggers/logidze_update.sql:
```diff
@@ -0,0 +1,20 @@
+CREATE TRIGGER logidze_before_update_on_<%= full_table_name %>
+BEFORE UPDATE ON <%= full_table_name %> FOR EACH ROW
```
Please, quote all the names with backticks as we cannot control `full_table_name`.
PG versions are quoted.

ardecvz commented on lib/generators/logidze/install/mysql/templates/migration.rb.erb:
Why are we omitting the "fx" version here? It appears that "fx" supports MySQL through a plugin.

Odebe replied:
I left this for later. The main priority was to make the migrations work and move on.

ardecvz commented on lib/generators/logidze/model/mysql/model_generator.rb:
```diff
@@ -0,0 +1,161 @@
+        class_option :update, type: :boolean, optional: true,
+          desc: "Define whether this is an update migration"
+
+        class_option :after_trigger, type: :boolean, optional: true,
```
Does it work for MySQL?
Does `only_trigger` work too?
Do we need to document `except` at all?

Odebe replied:
`only_trigger` option is used.
https://github.com/Odebe/logidze/blob/mysql_support/lib/generators/logidze/model/mysql/templates/migration.rb.erb#L3

> Do we need to document except at all?

It is a kind of indeterminate for me but I think this might be a thing for cases when someone calls mysql generator with postgresql arguments assuming they have same arguments.

Got it. I have to print warn about `after_trigger` not implemented for MySQL generator.

ardecvz commented on lib/generators/logidze/install/mysql/functions/logidze_compact_history.sql:
```diff
@@ -0,0 +1,32 @@
+      SET new_history_item = JSON_OBJECT(
+          'ts', JSON_EXTRACT(log_data, '$.h[1].ts'),
+          'v',  JSON_EXTRACT(log_data, '$.h[1].v'),
+          'c',  JSON_MERGE_PATCH(JSON_EXTRACT(log_data, '$.h[0].c'), JSON_EXTRACT(log_data, '$.h[1].c'))
```
Is it safe with NULL values?
Perhaps we should use this one, shouldn't we?
https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#json-merge-patch-json-merge-preserve-compared

Odebe replied:
> Is it safe with NULL values?

Kinda, it wouldn't raise exception, but will return null if second argument is null. 
PostgresQL version also returns null if second argument is null.
https://github.com/Odebe/logidze/blob/mysql_support/lib/generators/logidze/install/pg/functions/logidze_compact_history.sql#L13
```sql
postgres=# select ('{ "a": 1 }'::jsonb || null) as result;
 result
--------

(1 row)
```
> Perhaps we should use this one, shouldn't we?

I don't think so.
It appends values instead of merging. It will break diff calculation logic as minimum. 
```sql
mysql> SET @x = '{ "a": 1, "b": 2 }',
     >     @y = '{ "a": 3, "c": 4 }',
     >     @z = '{ "a": 5, "d": 6 }';

mysql> SELECT  JSON_MERGE_PATCH(@x, @y, @z)    AS Patch,
    ->         JSON_MERGE_PRESERVE(@x, @y, @z) AS Preserve\G
*************************** 1. row ***************************
   Patch: {"a": 5, "b": 2, "c": 4, "d": 6}
Preserve: {"a": [1, 3, 5], "b": 2, "c": 4, "d": 6}
```

So behaviour of `JSON_MERGE_PATCH(json, json)` is similar to ` jsonb || jsonb` in postgresql.
```sql
postgres=# select ('{ "a": 1 }'::jsonb || '{ "a": 2 }'::json) as result;
  result
----------
 {"a": 2}
(1 row)
```

If I need to take care of null in second argument then I would write something like:
```
(Pseudocode)
if '$.h[1].c' is not null then
  merge '$.h[1].c' into '$.h[0].c'
else
  return '$.h[0].c'
end
```

ardecvz replied:
Oh, I see now, but I mean another case:
we have 2 history entities - the first one is a full snapshot (a: 1, b: 2), and the second one is partial (a: 1, b: null).
So if we use JSON_MERGE_PATCH, we'll lose b: in the full snapshot after shrinking.
Or is it not true?

Odebe replied:
Yea, this is true.
Seems like right way to merge is iteration over keys with call `JSON_SET` for each key.

ardecvz commented on lib/generators/logidze/install/mysql/functions/logidze_logger.sql:
```diff
@@ -0,0 +1,81 @@
+    ELSE
+        SET log_data = row_log_data;
+
+        IF trigger_type = 'UPDATE' AND (old = new) THEN
```
My main big concern about this part - why do we make full snapshots **by design?**
Is there any real fundamental obstacles in MySQL?
It seems we can get all column values in OLD and compare them with changed only in NEW - you do something like this here.

Odebe replied:
I mistackenly put all snapshots inside unimplemented list. 
These is only difficulties in calling snapshot from ruby code. 
https://github.com/Odebe/logidze/blob/master/lib/logidze/model.rb#L58

But as I can understand now - difficulties only in requirement of specifying 'only' option for method call for logidze users.

Odebe committed:
DATABASE_RUL fallback for dummy

Odebe committed:
reusing migrations

Odebe committed:
removing Logidze::Utils::FunctionDefinitions module

Odebe committed:
typo

Odebe committed:
adding Trilogy support

Odebe committed:
contant renaming

Odebe committed:
escaping table name with backticks

Odebe committed:
after_trigger guard

Odebe committed:
adding fx support for mysql generators + model:install tests

Odebe committed:
proper attribute merging in logidze_compact_history

Odebe committed:
rubocop

Odebe committed:
mysql: snapshot

Odebe committed:
timestamp_column option warn

Odebe commented:
Kinda changelog:

1. Added `DATABASE_URL` fallback
2. Added concern for migrations reuse
3. Removed `Logidze::Utils::FunctionDefinitions` (with its clients refactoring)
4. `CantDefineDatabase` error renamed to `CannotIdentifyDatabas`
5. `load.rb` renamed to `current.rb`
6. Added warnings for `after_trigger`, `debounce_time` and `timestamp_column` options in `model:mysql:install`
7. Added FX support + `model:install` generator tests
8. Added table name escaping in sql (templates and execute blocks)
9. Implemented snapshots for MySQL
10. `JSON_MERGE_PATCH` replaced with loop

Odebe committed:
undefined method `[]` for ActiveRecord::DatabaseConfigurations::UrlConfig error fix

Odebe commented:
[Benchmarks ](https://github.com/Odebe/logidze/blob/mysql_benchmarks/bench/performance/mysql/README.md) [[PR](https://github.com/Odebe/logidze/pull/2)]

Odebe committed:
mysql: fixing columns_j double render

Odebe committed:
fixing Rails 7 deprecation warning

Odebe committed:
rubocop

palkan commented on spec/dummy/.byebug_history:
```diff
@@ -0,0 +1,59 @@
+exit
```
Oops
